File: src/VoidRay/Services/Brand.cs
using System.Text.RegularExpressions;

namespace VoidRay.Services;

/// <summary>Everything specific to the EnderrVPN service.</summary>
public static class Brand
{
    public const string ServiceName = "EnderrVPN";
    public const string SubHost = "sub.enderr.win";
    public const string LinkExample = "https://sub.enderr.win/ender/…";
    public const string DefaultSupportUrl = "https://discord.gg/BzG6FJz6zK";

    private static readonly Regex LinkPattern = new(
        @"^https://sub\.enderr\.win/ender/(?<sid>[A-Za-z0-9_-]{4,64})/?$",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

    /// <summary>Accepts only EnderrVPN subscription links and returns them normalised.</summary>
    public static bool TryNormalizeLink(string? input, out string link, out string sid)
    {
        link = "";
        sid = "";
        var m = LinkPattern.Match((input ?? "").Trim());
        if (!m.Success)
            return false;
        sid = m.Groups["sid"].Value;
        link = $"https://{SubHost}/ender/{sid}";
        return true;
    }
}

File: android/app/src/main/java/win/enderr/voidray/core/Brand.kt
package win.enderr.voidray.core

/** Everything specific to the EnderrVPN service. */
object Brand {
    const val SERVICE_NAME = "EnderrVPN"
    const val SUB_HOST = "sub.enderr.win"
    const val LINK_EXAMPLE = "https://sub.enderr.win/ender/…"
    const val DEFAULT_SUPPORT_URL = "https://discord.gg/BzG6FJz6zK"

    private val linkPattern = Regex(
        """^https://sub\.enderr\.win/ender/([A-Za-z0-9_-]{4,64})/?$""",
        RegexOption.IGNORE_CASE,
    )

    /** Accepts only EnderrVPN subscription links; returns (normalised link, subscription id). */
    fun normalizeLink(input: String?): Pair<String, String>? {
        val m = linkPattern.matchEntire(input.orEmpty().trim()) ?: return null
        val sid = m.groupValues[1]
        return "https://$SUB_HOST/ender/$sid" to sid
    }
}
